### DISCUSSION TITLE
Can't insert rows into database with `@vercel/postgres`

### DISCUSSION THREAD

**Discussion by KovJonas** [vercel/community#2336] [Help]
### Summary

Hey everyone! 👋 

I'm playing around with the fantastic new Vercel storage features, and it seems like I just can't insert data into my database using `@vercel/postgres` with an on-the-fly generated sql query, I believe I'm missing something here. If I `console.log` the generated query, and manually run it on Vercel, it works as expected. It also works if I paste the logged query into my source code and run it, but it doesn't work if it's generated on-the-fly. 👀 Am I doing something wrong?

Here's an example on how I generate it:

```ts
const myArray = [
  {
    name: "Something",
    image: "https://...",
    price: 9.99
  },
  {
    name: "Anything",
    image: "https://...",
    price: 5.99
  }
]

const { rows } = await sql`insert into table_name (col1, col2, col3) values ${myArray.map((item) => `(${item.name}, ${item.image}, ${item.price})`).join()}  returning *;`
```

### Additional information

Important note:
- I'm using `@vercel/postgres` in a simple `.ts` file which I build using `esbuild` and then run it with `node` - because I'll run it inside a cron job / GitHub Action. I also run `select` queries in the same file and those work properly.

Package versions:
- `@vercel/postgres`: 0.1.3

Error message:

```
let a2 = s3.M, u2 = i3 === "notice" ? new j.NoticeMessage(t2, a2) : new j.DatabaseError(a2, t2, i3);
                                                                    ^
error: syntax error at or near "$1"
// some trace here...
{
  length: 90,
  severity: 'ERROR',
  code: '42601',
  detail: undefined,
  hint: undefined,
  position: '1',
  internalPosition: undefined,
  internalQuery: undefined,
  where: undefined,
  schema: undefined,
  table: undefined,
  column: undefined,
  dataType: undefined,
  constraint: undefined,
  file: 'scan.l',
  line: '1188',
  routine: 'scanner_yyerror'
}
```

**Comment by dwhitegithub** [comment]
I have a similar problem in that inserts - both singleton and bulk - succeed from an application run on localhost, but the same code published to vercel, singleton inserts succeed but bulk inserts fail (same data in all cases).

**Comment by mooky1007** [comment]
Hi, I had a similar problem.
I think it's an error in the grammar of tagged templates.

It seems to be an issue that arises because there is another template literature inside the template literature.

https://developer.mozilla.org/ko/docs/Web/JavaScript/Reference/Template_literals#tagged_templates

**Comment by basit-sharif** [comment]
Can you please run this query as the map need a curly braces :
`const { rows } = await sql`insert into table_name (col1, col2, col3) values ${myArray.map((item) => {return`(${item.name}, ${item.image}, ${item.price})`}).join()}  returning *;``

**Comment by basit-sharif** [comment]
The map needs to be return something into like braces
`const { rows } = await sql`insert into table_name (col1, col2, col3) values ${myArray.map((item) => {return`(${item.name}, ${item.image}, ${item.price})`)}.join()}  returning *;`
`

**Reply by mooky1007** [reply]
hello

try this code.

```js
const test = (string, ...values) => {
    console.log(string);
    console.log(values);
}

const testName = 'jhon';
const testTemplate = `hello ${testName}`

test`hello ${testName}`
test`${testTemplate}`
```

do not use nesting

**Comment by basit-sharif** [comment]
What happened?

**Reply by mooky1007** [reply]
sql\`${somting}\` -> not working
sql\`text ${somting}\` -> working

check console

your code.
```js
myArray.forEach(item => {
    sql`insert into table_name (col1, col2, col3)  values(${item.name}, ${item.image}, ${item.price});
})
```


This is sample code.
To actually use it, it must be processed using Promise All, etc.

**Comment by basit-sharif** [comment]
I didn’t understood why I am receiving this email?

**Comment by mikeyamato** [comment]
The solution that worked for me came via [SO](https://stackoverflow.com/a/77093839/8378893). `sql.query()` is what I needed in order to allow `Array.map()` to work. 

```javascript
await sql.query(
  `INSERT INTO table_name (col_a, col_b, col_c) VALUES ${items.map((item) => {
    return `('${item.a}', ${item.b}, '${item.c}')`;
  }).join()}`
);
```

* for example sake items `a` and `c` are strings while `b` is an int

**Comment by basit-sharif** [comment]
It works!
